File: PassedProblems/228-summary-ranges.cpp
/*
 * -summary-ranges.cpp
 *Given a sorted integer array without duplicates, return the summary of its ranges.

For example, given [0,1,2,4,5,7], return ["0->2","4->5","7"].

 *
 *  Created on: Sep 25, 2015
 *      Author: hanzhe
 */

#include <iostream>
#include <vector>
#include <sstream>
using namespace std;




class Solution {
public:
    vector<string> summaryRanges(vector<int>& nums) {
        vector<string> re;
        if(nums.size() < 1)
        	return re;
        int lb = nums[0], le = nums[0];
        for(int i = 1 ; i < nums.size(); i ++){
        	if(nums[i] == le + 1)
        		le ++;
        	else{
        		re.push_back(merge(lb, le));
        		lb = le = nums[i];
        	}
        }
        re.push_back(merge(lb,le));
        return re;
    }

    string merge(int lb, int le){
		stringstream ss;
		if(lb != le)
			ss << lb << "->" << le;
		else ss << lb;
		return ss.str();
    }
};

File: leetcode/python/228-summary-ranges.py
#!/usr/bin/env python
# encoding: utf-8



"""
Given a sorted integer array without duplicates, return the summary of its ranges.

For example, given [0,1,2,4,5,7], return ["0->2","4->5","7"].
"""


# 1次过
class Solution(object):
    def summaryRanges(self, nums):
        """
        :type nums: List[int]
        :rtype: List[str]
        """
        nl = len(nums)
        if nl == 0:return []
        ls, le = nums[0], nums[0]
        re = []
        def p(ls, le):
            if ls == le: re.append(str(ls))
            else: re.append(str(ls) + "->" + str(le))
        for i in range(1, nl):
            if nums[i] == le + 1: le += 1
            else:
                p(ls, le)
                ls = le = nums[i]
        p(ls, le)
        return re


